src/torrent.cpp

```diff
@@ -7429,23 +7429,52 @@ bool is_downloading_state(int const st)
 		if (lhs->on_parole() != rhs->on_parole())
 			return lhs->on_parole();
 
-		// prefer to disconnect peers that send data at a lower rate
+		// prefer to disconnect peers that send data at a lower rate over those waiting to download
+		// the assumption is that a connection waiting to start downloading has the same potential 
+		// to be faster than the current slowest downloading connection than a new connection that 
+		// replaces a waiting connection - and disconnecting a waiting connection starts a new wait
+		// however ... to avoid torrents which wait forever there needs to be a maximum wait time
+		// after which a waiting connection is disconnected
 		std::int64_t lhs_transferred = lhs->statistics().total_payload_download();
 		std::int64_t rhs_transferred = rhs->statistics().total_payload_download();
+		bool const is_waiting = lhs_transferred * rhs_transferred == 0;
 
+		// for this commit, the maximum time is a constant - but it could be replaced by a new 
+		// setting peer_turnover_maximum_wait in a future commit
+		std::int64_t const peer_turnover_maximum_wait = 60;
 		time_point const now = aux::time_now();
 		std::int64_t const lhs_time_connected = total_seconds(now - lhs->connected_time());
 		std::int64_t const rhs_time_connected = total_seconds(now - rhs->connected_time());
+		
+		// if one connection is waiting
+		if ((lhs_transferred != rhs_transferred) && (is_waiting))
+		{
+			// One of lhs_transferred & rhs_transferred is zero - get connected time for waiting connection
+			std::int64_t const waiting_time = 
+				((lhs_transferred * rhs_time_connected) + (rhs_transferred * lhs_time_connected)) 
+				/
+				(lhs_transferred + rhs_transferred);
+			if (waiting_time <= peer_turnover_maximum_wait)
+				return lhs_transferred != 0;
+			else return lhs_transferred == 0;
```

Comment:
if I understand this correctly, a "waiting" peer that connected < 60s ago is prioritized over a peer that's currently transferring in this case, right?

Clearly to get here the peers both need to be equal in a few other comparisons that happen first, like being a seed and interesting.

It seems like there's an edge case that could be problematic though. Imagine a swarm if only seeds. You join and immediately fill all your peer connection slots with seeds. One of them starts sending data but the other ones stall for some reasons.

If the optimistic disconnect logic is invoked within 60 seconds, the one transferring peer will be the top disconnect candidate and be disconnected.

I suppose that constant, `peer_turnover_maximum_wait` critically depends on how often the disconnect logic is invoked. I think there should be (at least) a 2x margin between those constants. And there should either be comment explaining that these constants are linked, or better yet, use the same constant.

Reply:
I will need to check, but you are right - we should NOT prefer a connection which is waiting over one that has only just started downloading. So yes - we should prioritise the downloading peer for (say) twice the peer_turnover_maximum_wait,

Reply:
On reflection I am not sure that 2x maximum wait is long enough to allow a connection to reach its maximum speed.

Perhaps we should use 3x instead.